### PR TITLE
[WIP] transducer: support operator| syntax for composition

### PR DESCRIPTION
**Pull request by CJBussey**
Here's an early sneak peak of the approach for making composition possible via operator|. Unfortunately, because lambda support for this sort of stuff is terrible in C++14, we need to convert some of the inner lambdas to structs. This is to ensure that the user code can do something like:

```
ZUG_INLINE_CONSTEXPR auto filter_lt10 = filter(lt10);
```

It's required to be a struct because lamdas can't be used in constant expressions before C++17, so initializing a constexpr variable from `make_transducer_holder(lambda)` doesn't work.

I haven't moved all of them over to structs yet - the other ones are just using `make_transducer_holder(lambda)`, but aren't used in a constexpr.

Let me know if you agree with the general approach (before I attack everything!).

### CODE REVIEW

File: zug/transducer/map_indexed.hpp [hunk 1]
```diff
@@ -19,7 +20,7 @@ namespace zug {
 template <typename MappingT>
 constexpr auto map_indexed(MappingT&& mapping)
 {
-    return comp(count(), map(std::forward<MappingT>(mapping)));
+    return detail::make_transducer_holder(comp(count(), map(std::forward<MappingT>(mapping))));
```
Comment (CJBussey) [inline zug/transducer/map_indexed.hpp]: Ignore these changes... The changes to the composed transducers won't be needed.

**Comment by arximboldi** [comment]
Hey!

It is moving in the right direction, thanks a lot! 

A couple of comments:

1. I do not like turning all transducers into structs. So far, I was doing it only for transducers that have no factory function. Why did you do the change? I do not see why introducing this new feature changes things.

2. I do not like the name `transducer_holder`, i'd rather have it be `pipeable` or `composable`. Also, maybe we can mix that with the result type from `comp(...)` (this `composed`). I have been thing of reimplementing the latter for some time actually, to make it a bit more efficient, inheriting from a `std::tuple<...>` that holds all the functions. In this way empty functions do not occupy space. By implementing `operator|` and having the base case be only function, this would serve the purpose of enabling pipeability too.

3. It seems `operator|` has only be implemented for r-value references? Is there some rationale for that? To avoid combinatorial explossion, I would just implement it for value types and use `std::move` inside.

**Comment by CJBussey** [comment]
> Hey!
> 
> It is moving in the right direction, thanks a lot!
> 
> A couple of comments:
> 
> 1. I do not like turning all transducers into structs. So far, I was doing it only for transducers that have no factory function. Why did you do the change? I do not see why introducing this new feature changes things.

The most obvious example is `enumerate_from`. Constructing a `ZUG_INLINE_CONTEXPR auto enumerate` from `enumerate_from` requires a constant expression. C++14 doesn't support lambdas in constant expressions, so the `make_transducer_holder([...](...){...})` doesn't work. The workaround is to move it to a struct.

Whether we want to do this for the rest of the transducers is a question of how it will be used. Since I didn't want to make an assumption there, I thought it would be best to move them all. What do you think?

> 2. I do not like the name `transducer_holder`, i'd rather have it be `pipeable` or `composable`. Also, maybe we can mix that with the result type from `comp(...)` (this `composed`). I have been thing of reimplementing the latter for some time actually, to make it a bit more efficient, inheriting from a `std::tuple<...>` that holds all the functions. In this way empty functions do not occupy space. By implementing `operator|` and having the base case be only function, this would serve the purpose of enabling pipeability too.

Let me think about that a bit more...

> 3. It seems `operator|` has only be implemented for r-value references? Is there some rationale for that? To avoid combinatorial explossion, I would just implement it for value types and use `std::move` inside.

Yes, that was a mistake. I was initially expressing forwarding references, but after some changes, it turned into this. I like the idea of using values.

**Comment by arximboldi** [comment]
> The most obvious example is enumerate_from. Constructing a ZUG_INLINE_CONTEXPR auto enumerate from enumerate_from requires a constant expression. C++14 doesn't support lambdas in constant expressions, so the make_transducer_holder([...](...){...}) doesn't work. The workaround is to move it to a struct.

Why was it compiling before?

**Comment by CJBussey** [comment]
Because we didn’t use the lambda in a constant expression. That is, we
didn’t pass the lambda into a function and expect the function to be then
constexpr. Initializing a constexpr lambda isn’t a problem with C++14, but
using it in a constant expression is.

**Comment by arximboldi** [comment]
> Because we didn’t use the lambda in a constant expression. That is, we
> didn’t pass the lambda into a function and expect the function to be then
> constexpr. Initializing a constexpr lambda isn’t a problem with C++14, but
> using it in a constant expression is.

And what is the constexpr function now, the `operator|`? Can we have
constexpr and non-constexpr versions?  I still do not understand what is
the new constexpr use... it seems we are just putting a wrapper around
the previously existing lambda...

I am still hesitant about going all in into making the library constexpr
friendly and would prefer to keep using lambdas as much as
possible... Maybe it is a stupid obsession of mine, but would prefer to
keep changes minimal for now until I give it more thought.

Cheers!

  JP

-- 
∿∿∿  https://sinusoid.al
∿∿∿  https://sinusoid.es

**Comment by CJBussey** [comment]
For documentation purposes, here's a summary of our offline discussion. The problem is demonstrated minimally here:

https://godbolt.org/z/oBowyx

On older compilers, the compiler complains at L50 only. With newer compilers, it already complains at L25.

As discussed, the way to go forward would be to only move those that need to be moved to structs to structs (i.e., `enumerate`).

**Pull request closed without merging**